File: src/main/java/com/rdvdev2/timetravelmod/impl/common/timemachine/TimeMachineManager.java
package com.rdvdev2.timetravelmod.impl.common.timemachine;

import com.rdvdev2.timetravelmod.impl.ModRegistries;
import net.minecraft.block.BlockState;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import org.apache.commons.lang3.ArrayUtils;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public class TimeMachineManager {

    private static final TimeMachineManager INSTANCE = new TimeMachineManager();

    public static TimeMachineManager getInstance() {
        return INSTANCE;
    }

    private final Map<BlockState, TimeMachineStructure[]> possibleStructures = new ConcurrentHashMap<>();

    public void generateStructures() {
        possibleStructures.clear();

        ModRegistries.TIME_MACHINE.stream().parallel().forEach(tm -> {
            BlockState[] controllerStates = tm.getControllerStates();
            TimeMachineStructure[] structures = TimeMachineStructure.generateFromTimeMachine(tm);
            for (BlockState controllerState : controllerStates) {
                TimeMachineStructure[] all = ArrayUtils.addAll(possibleStructures.getOrDefault(controllerState, null), structures);
                possibleStructures.put(controllerState, all);
            }
        });
    }

    public Optional<TimeMachineExecutor> generateExecutor(World world, BlockPos rootPos) {
        BlockState rootState = world.getBlockState(rootPos);
        if (!possibleStructures.containsKey(rootState)) generateStructures();
        if (!possibleStructures.containsKey(rootState)) return Optional.empty();

        for (TimeMachineStructure structure: possibleStructures.get(rootState))
            if (structure.checkIfBuilt(world, rootPos)) return Optional.of(new TimeMachineExecutor(structure, rootPos, world));
        return Optional.empty();
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/impl/common/timemachine/TimeMachineManager.kt
package com.rdvdev2.timetravelmod.impl.common.timemachine

import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachine
import com.rdvdev2.timetravelmod.impl.ModRegistries
import net.minecraft.block.BlockState
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import org.apache.commons.lang3.ArrayUtils
import java.util.*
import java.util.concurrent.ConcurrentHashMap

object TimeMachineManager {

    private val possibleStructures: MutableMap<BlockState, Array<TimeMachineStructure>> = ConcurrentHashMap()

    private fun generateStructures() {
        possibleStructures.clear()
        ModRegistries.TIME_MACHINE.stream().parallel().forEach { tm: ITimeMachine ->
            val controllerStates = tm.controllerStates
            val structures = TimeMachineStructure.generateFromTimeMachine(tm)
            for (controllerState in controllerStates) {
                val all = ArrayUtils.addAll(possibleStructures.getOrDefault(controllerState, null), *structures)
                possibleStructures[controllerState] = all
            }
        }
    }

    fun generateExecutor(world: World, rootPos: BlockPos): Optional<TimeMachineExecutor> {
        val rootState = world.getBlockState(rootPos)
        if (!possibleStructures.containsKey(rootState)) generateStructures()
        possibleStructures[rootState]?.forEach { structure ->
            if (structure.checkIfBuilt(world, rootPos)) return Optional.of(TimeMachineExecutor(structure, rootPos, world))
        }
        return Optional.empty()
    }
}